components/uniqueness/backing-store-impl/src/integrationTest/kotlin/net/corda/uniqueness/backingstore/impl/JPABackingStoreImplIntegrationTests.kt

```diff
@@ -0,0 +1,487 @@
+import javax.persistence.EntityManagerFactory
+import javax.persistence.OptimisticLockException
+import javax.persistence.PersistenceException
+import javax.persistence.QueryTimeoutException
+import javax.persistence.RollbackException
+import javax.persistence.EntityNotFoundException
+import javax.persistence.LockTimeoutException
+import javax.persistence.NoResultException
+import javax.persistence.NonUniqueResultException
+import javax.persistence.TransactionRequiredException
+import javax.persistence.PessimisticLockException
+import kotlin.test.assertEquals
+import kotlin.test.assertTrue
+
+/**
+ * Hint: To run tests against PostgreSQL, follow the steps in the link below.
+ * https://github.com/corda/corda-runtime-os/wiki/Debugging-integration-tests#debugging-integration-tests-with-postgres
+ */
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class JPABackingStoreImplIntegrationTests {
+    private val maxAttemptsCnt = 10 // Set it same as "MAX_ATTEMPTS" in JPABackingStoreImpl.kt
+    private lateinit var backingStoreImpl: JPABackingStoreImpl
+
+    private val entityManagerFactory: EntityManagerFactory
+    private val dbConfig = DbUtils.getEntityManagerConfiguration("uniqueness_default")
+
+    private lateinit var lifecycleCoordinator: LifecycleCoordinator
+    private lateinit var lifecycleCoordinatorFactory: LifecycleCoordinatorFactory
+
+    class DummyException(message: String) : Exception(message)
+
+    init {
+        entityManagerFactory = createEntityManagerFactory("uniqueness_default")
+    }
+
+    @BeforeEach
+    fun init() {
+        lifecycleCoordinator = mock<LifecycleCoordinator>()
+        lifecycleCoordinatorFactory = mock<LifecycleCoordinatorFactory>().apply {
+            whenever(createCoordinator(any(), any())) doReturn lifecycleCoordinator
+        }
+
+        backingStoreImpl = createBackingStoreImpl(entityManagerFactory)
+        backingStoreImpl.eventHandler(RegistrationStatusChangeEvent(mock(), LifecycleStatus.UP), mock())
+    }
+
+    private fun createBackingStoreImpl(emFactory: EntityManagerFactory): JPABackingStoreImpl {
+        return JPABackingStoreImpl(lifecycleCoordinatorFactory,
+            JpaEntitiesRegistryImpl(),
+            mock<DbConnectionManager>().apply {
+                whenever(getOrCreateEntityManagerFactory(any(), any(), any())) doReturn emFactory
+                whenever(getClusterDataSource()) doReturn dbConfig.dataSource
+            })
+    }
+
+    private fun createEntityManagerFactory(persistenceUnitName: String): EntityManagerFactory {
+        return EntityManagerFactoryFactoryImpl().create(
+            persistenceUnitName, JPABackingStoreEntities.classes.toList(), dbConfig
+        )
+    }
+
+    private fun generateSecureHashes(cnt: Int): LinkedList<SecureHash> {
+        val secureHashes = LinkedList<SecureHash>()
+        repeat(cnt) {
+            secureHashes.push(SecureHashUtils.randomSecureHash())
+        }
+        return secureHashes
+    }
```

Comment:
```suggestion
    private fun generateSecureHashes(cnt: Int) = List(cnt) { SecureHashUtils.randomSecureHash() }
```

Could be simplified to the above. This is then sufficiently trivial that I'd question whether it's worth a dedicated function, or if we just inline the above in your tests.

If we do want a separate function, I'd suggest moving it to a common `listOfRandomSecureHashes` function in `SecureHashUtils`, as this function is duplicated in the unit tests.

Reply:
- [ ] Remove generateSecureHashes with an inline code. 
- [ ] Review and update use of LinkedList(). Replace with List().